libparsec/crates/client/tests/unit/certif/validate_block.rs

```diff
@@ -0,0 +1,428 @@
+            data.wksp1_last_key_index,
+            &data.wksp1_bar_txt_manifest,
+            &data.wksp1_bar_txt_block_access,
+            &data.wksp1_bar_txt_block_encrypted,
+        )
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        CertifValidateBlockError::InvalidBlockAccess(boxed)
+        if matches!(*boxed, InvalidBlockAccessError::HashDigestMismatch { .. })
+    );
+}
+
+#[parsec_test(testbed = "minimal_client_ready")]
+async fn invalid_block_access_cannot_decrypt(env: &TestbedEnv) {
+    let data = BaseData::extract(env);
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    test_register_send_hook(
+        &env.discriminant_dir,
+        test_send_hook_realm_get_keys_bundle!(env, alice.user_id, data.wksp1_id),
+    );
+
+    let err = ops
+        .validate_block(
+            data.wksp1_last_realm_certificate_timestamp,
+            data.wksp1_id,
+            data.wksp1_last_key_index,
+            &data.wksp1_bar_txt_manifest,
+            &data.wksp1_bar_txt_block_access,
+            b"<dummy encrypted>".as_ref(),
+        )
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        CertifValidateBlockError::InvalidBlockAccess(boxed)
+        if matches!(*boxed, InvalidBlockAccessError::CannotDecrypt { .. })
+    );
+}
+
+#[parsec_test(testbed = "minimal_client_ready")]
+async fn invalid_block_access_non_existent_key_index(env: &TestbedEnv) {
+    let data = BaseData::extract(env);
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    test_register_send_hook(
+        &env.discriminant_dir,
+        test_send_hook_realm_get_keys_bundle!(env, alice.user_id, data.wksp1_id),
+    );
+
+    let err = ops
+        .validate_block(
+            data.wksp1_last_realm_certificate_timestamp,
+            data.wksp1_id,
+            42,
+            &data.wksp1_bar_txt_manifest,
+            &data.wksp1_bar_txt_block_access,
+            &data.wksp1_bar_txt_block_encrypted,
+        )
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        CertifValidateBlockError::InvalidBlockAccess(boxed)
+        if matches!(*boxed, InvalidBlockAccessError::NonExistentKeyIndex { .. })
+    );
+}
+
+#[parsec_test(testbed = "minimal_client_ready")]
+async fn invalid_block_access_corrupted_key(env: &TestbedEnv) {
+    let wksp1_id: VlobID = *env.template.get_stuff("wksp1_id");
+    // We do another key rotation to be able to provide our own keys bundle containing
+    // a corrupted key for key index 1 (i.e. the key used to encrypt the block).
+    env.customize(|builder| {
+        builder.rotate_key_realm(wksp1_id).customize(|e| {
+            p_assert_eq!(e.keys.len(), 2); // Sanity check
+                                           // Corrupt key with index 1
+            e.keys[0] = KeyDerivation::generate();
```

Comment:
```suggestion
            // Corrupt key with index 1
            e.keys[0] = KeyDerivation::generate();
```

Reply:
`cargo fmt` got messed up by the comment on the previous line :/

fixed